gapic/templates/tests/unit/gapic/%name_%version/%sub/test_%service.py.j2

```diff
@@ -2459,4 +2465,51 @@ async def test_test_iam_permissions_from_dict_async():
 
 {% endif %}
 
+def test_transport_ctx():
+    transports = [
+        {% if 'rest' in opts.transport %}
+        'rest',
+        {% endif %}
+        {% if 'grpc' in opts.transport %}
+        'grpc',
+        {% endif %}
+    ]
+    for transport in transports:
+        client = {{ service.client_name }}(
+            credentials=ga_credentials.AnonymousCredentials(),
+            transport=transport
+        )
+        {% if 'rest' in opts.transport %}
+        with mock.patch.object(type(client.transport._session), 'close') as close:
+        {% endif %}
+        {% if 'grpc' in opts.transport %}
+        with mock.patch.object(type(client.transport._grpc_channel), 'close') as close:
+        {% endif %}
```

Comment:
My bad, I may have screwed this up in my endeavor to generalize the test. I don't think this is quite right now; we want to patch the object based on what type it is, not just on what the valid transports are. Both can be valid during the same generator run. We can probably fix it like so:
```python
transports = {
    {% conditions, don't want to retype %}
    "rest": "_session",
    {% conditions, don't want to retype %}
    "grpc": "_grpc_channel",
}

for transport, close_name in transports.items():
    client  = ...
    with mock.patch.object(type(getattr(client.transport, close_name)), "close") as close:
        ...
```

But that feels like it may be trying to be too clever.
One alternative is to go back to the previous change, and just duplicate the majority of the template, but I'm not hugely happy with that either. What do you think?

Reply:
I like using `dict` here since it doesn't tangle test function code (unlike the `array` approach), and also there is less redundant code (from two different tests).